### ISSUE TITLE
Turn occupation type into a customizable enum [5]

### ISSUE THREAD

**Issue by carolinverset** [hzi-braunschweig/SORMAS-Project#5015]
### Situation Description
The variable type of occupation (case/contact) doesn't seem suitable for monitoring all pathologies and deserves to be reworked.

### Feature Description
We propose to make this list configurable via a database table.
To manage the existing one, we start with an automatic creation of the table with the professions which are currently hard in the code and a script which migrates the data of the persons.
We provide:
- the script for adding a profession
- a script for deleting a profession
- a script to check if a profession is currently used (to know if we can delete)

### Additional Information
@bernardsilenou @ftavin @emond-fuseau

**Comment by MateStrysewske** [comment]
Please refer to my comment in #5014 here as well

**Comment by bernardsilenou** [comment]
@MateStrysewske Perfect !
So how do you suggest we go about this configuration issue please? 
Can each team do a review to come up with the optimum solution?

**Comment by MateStrysewske** [comment]
@carolinverset @ftavin We had a lengthy discussion about this topic yesterday and came up with a concept that we would like to be used going forward for anything related to configurable enum values. It would involve having a `CustomEnum` class that is extended by any current (or future) enum-like "thing" that needs to be customised, so in this case `OccupationType` would be removed as an enum and instead be transformed into a class that extends CustomEnum. There would be a `customenum` database table that contains all custom enum values of all classes (e.g. `diseasevariant` would be removed and migrated to that table at some point as well).

No matter which approach we take here, this issue is pretty complex because there's custom logic involved with occupation types, e.g. there are "occupation facility" fields and an "occupation details" field that are only visible for specific occupation types, and the "occupation details" field has a different caption based on the selected occupation type. We still need that logic, but making the occupation type configurable would mean that it can no longer be directly hard-coded (as in if occupation type = VALUE, then show the field with that caption).

With the approach we have in mind, there would be additional configurations stored as json in the database which would then be added as properties to the extending class (OccupationType in this case), e.g. "hasFacilityDetails", "hasOccupationDetails", "occupationDetailsCaption". That would allow us to preserve that logic.

In any way, we would like such a system to be in place before we go on with customisable enums. Would you be willing to implement such a system based on detailed implementation instructions I would forward to you? Otherwise I would try to prioritise this internally such as that I can implement it relatively soon so that you can then use that new system for your development.

**Comment by aponcon-atolcd** [comment]
Hi @MateStrysewske 
Having a solution that avoid multiplying the number of tables is a good idea indeed. But we need to be careful to not make the system too complicated.
If the solution will require a lot of work for something that will hardly fit with all our needs then I think that's not worth it.

I mean I prefer adding a lot of tables (which is very easy to do, easy to read and to extend if necessary) than developing a complex system based on configuration files.
I understand it will make adding new basic enum entities as simple as adding a new line in the configuration file.
But each time we will have a different use case, it will require us to extend this system, increasing heavily the amount of time required to develop it.

As an example, I've already worked on several softwares based on a lot of configuration files to make everything generic and it finally always made everything more complicated :/

**Comment by MateStrysewske** [comment]
@aponcon-atolcd The basic idea is that there won't be any configuration files, but one single database table that contains all the enum definitions. Long-term, those should also be configurable from the user interface (which of course is totally out of scope for now). I think the system wouldn't actually be all that complicated:

* Instead of a Java enum, we would use a new class that extends `CustomEnum`. This class would contain all the enum values, and those again will have a number of properties:
  * A caption in the main language of the server
  * A list of translations for other languages used on that server
  * A list of properties that define additional logic, e.g. whether specific fields would be toggled by this enum value; however, this would be very explicit, so you wouldn't specify the names of the fields here, but simply "hasFacilityDetails", which would then toggle the visibility of the "occupation facility" fields
  * Some additional properties that aren't relevant for now
* During server startup, we would build a cache that contains all the enum values for all the custom enum classes
* In the UI, dropdowns would be filled based on that cache
* Logic in the UI would be based on the additional properties in the individual enum values; this would replace the current logic that is there anyway, but it would actually move the definition of that logic out of the UI into the API which would be pretty nice

Adding custom enum values would then just mean to add another row to the customenums table; in its most basic state (i.e. no additional logic required), it would only include the enum class name, the enum value and its caption. Nothing more would be mandatory.

Setting up such a system would require some development work, but I think it's actually not that complex - and it's a lot cleaner than having a lot of extra database tables for all the enums that need to be configurable (which will be A LOT long-term).

**Comment by ftavin** [comment]
@MateStrysewske 

I agree with @aponcon-atolcd (I had similar bad experiences in earlier projects) and I think this will add a lot of complexity to the code itself, from enums to mechanisms such as imports/exports. The use of additional logic in a column would, I believe, make it even more complicated.

I think this design matches the 5th point of this article : https://www.red-gate.com/simple-talk/sql/database-administration/ten-common-database-design-mistakes/ but in an even more complicated way (again, for me, the addition of logic in columns).

Not to mention that you would move a lot of your logic from the (static) code to the (dynamic) database which means that every instance will be able to modify it and assistance will be a real pain for developpers (is this a bug or just an error in the magic table logic ? you need a table dump to check...).

That's why I truly believe that many dedicated tables is way more cleaner than a unique, incomprehensible, table.

**Comment by MateStrysewske** [comment]
@ftavin I think there are some misunderstandings here :) our suggestion is not to move logic into the database table, but simply properties that define which logic is executed. The actual logic would still be in the UI of course. If you want to introduce custom enum values - which you have to do anyway - you necessarily have to move the definition of which logic is executed to a generic place, there's no way around it (even if you define a single database table for every enum).

So right now, the code specifically states that for enum value X, field Y should be visible. The generic approach would be to say if enum value X has property Z, field Y should be visible. So it actually moves the definition out of the UI and into the API which is pretty nice, but it of course keeps the logic itself in the UI.

Maybe it would be good to have a call about this topic because I think the topic is a bit too complex to discuss it in written form right here?

**Comment by bernardsilenou** [comment]
@fta @MateStrysewske What about using the next refinement meeting for this? 
@MateStrysewske can you be available form 11:00 to 12:00 on Wednesday 28 April?

**Comment by MateStrysewske** [comment]
@bernardsilenou I don't know whether it makes a lot of sense for you or someone else from HZI to be part of this discussion because it's a very technical one. Also the 28th of April might be a bit late depending on how urgent this issue is for the French team?

**Comment by bernardsilenou** [comment]
@MateStrysewske I meant 21, its was a mistake in typing. I suggested this day because its a regular meeting we always have. We cancel the idea and if needed you can negotiate with @ftavin for another date.

**Comment by MateStrysewske** [comment]
Quick note: We should probably define "OTHER" and "UNKNOWN" as default values that can be toggled on/off in the customizable enum (in the code).

**Comment by MateStrysewske** [comment]
@JaquM-HZI @kwa20 We still need default values here that will be present on each server.

**Comment by kwa20** [comment]
@MateStrysewske I assume those would include everything that needs to be migrated, including the initially wrongfully translated values which then need to become their own additional values?

**Comment by MateStrysewske** [comment]
@kwa20 I think the initially wrongfully translated values need to be fixed separately as they don't have anything to do with SORMAS itself, but are a problem of the German instances. What we need is basically a list of the least common ground of occupations for all countries, which individual countries can then build upon. We will then ensure that any running server still retains the occupation types that they have been using so far of course in order to not break anything.

**Comment by kwa20** [comment]
@MateStrysewske Another thing that came to mind is that we probably need an archived column for customizable enum values to make entries unavailable if they're no longer applicable without changing existing entries.

**Comment by MateStrysewske** [comment]
@kwa20 The `value` column of customizable enums is basically an enum value, so whatever occupation types are currently in use on a system will be added there in order to ensure compatibility, and the `caption` column will be filled with the translation based on the server language (we might as well fill the `translations` column with all the additional translations, though I'm not sure whether that's really necessary and warrants the effort).

For default values that we provide, those will indeed be translatable via Crowdin, i.e. they will still be added to the database so that server admins can e.g. delete them if they don't want to use them, but they won't have to provide translations for them themselves. Probably makes sense to add a "defaultValue" column to the table in order to hint at the value being pre-configured. Custom values created by server admins will of course not be translated via Crowdin.

And yes, I agree that we need such a column (would call it `active` with default value true I think), but that should be tackled in a separate issue.

**Comment by kwa20** [comment]
As discussed with @JaquM-HZI and @Jan-Boehme, the minimal possible disease independent values as default occupation types should merely be:

default values | Caption
-- | --
HEALTHCARE_WORKER | Healthcare worker
LABORATORY_STAFF | Laboratory staff
OTHER | Other
UNKNOWN | Unknown

All existing SORMAS users will retain their previously used types and new users will be able to customize their values based on the diseases they have enabled.

**Comment by adinaflorea9** [comment]
Ticket is still in the testing lane as I can not test it on my local machine as the changes have already been pulled from development. Prepared the test data on release-international.sormas.netzlink.com and once the release of 1.75.0 is done I will be able to validate the ticket there.

I attempted multiple times to revert my local machine to an earlier version of SORMAS  or to at least an earlier commit but the server does not start.

**Comment by adinaflorea9** [comment]
Verified ticket on the local machine using the latest version of SORMAS from the development branch - 1.75.0-SNAPSHOT(a003252).
